exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java

```diff
@@ -1221,6 +1223,12 @@ public void interrupted(final InterruptedException e) {
      *   to the user
      */
     public void moveToState(final QueryState newState, final Exception ex) {
+      // if the current thread is the foreman thread, throw an exception
+      // otherwise the foreman will be blocked forever on acceptExternalEvents
+      if (myThreadRef == Thread.currentThread()) {
```

Comment:
We were assuming that the Foreman thread would never call Foreman.StateListener.moveToState() and it would be called by another (rpc) thread instead.
It turns out when the foreman is submitting remote fragments, RpcBus.send() could actually cause the foreman thread to call moveToState directly